### PR TITLE
ctop: top-like interface for container metrics

### PR DESCRIPTION
**Pull request by jmarcet**
Maintainer: Javier Marcet <javier@marcet.info>
Compile tested: x86-64
Run tested: x86-64

Description:

new package ctop: top-like interface for container metrics

### CODE REVIEW

**Comment by jmarcet** [comment]
Does anyone know what problem there is building go on mips? I guess I can add a DEPENDS which blocks mips architecture.

**Comment by neheb** [comment]
Sounds like you need PKG_USE_MIPS16:=0

**Comment by jmarcet** [comment]
Updated to stable release and added PKG_USE_MIPS16:=0

It still fails CI on mips.

**Comment by jmarcet** [comment]
Instead of PKG_USE_MIPS16:=0 I added @mips to DEPENDS, now CI is ok.

**Comment by neheb** [comment]
Why is that? The pattern for other go packages is to use PKG_USE_MIPS16:=0

**Comment by jmarcet** [comment]
> Why is that? The pattern for other go packages is to use PKG_USE_MIPS16:=0

Well, with PKG_USE_MIPS16:= it sill failed CI, that's why. I can revert it if you deem it worth.

**Comment by neheb** [comment]
What was the failure?

**Comment by jmarcet** [comment]
It was a build failure, some opcodes not available on mips. I'd need to revert the change to see it, I don't have a copy of the error.

**Comment by neheb** [comment]
This is the error:

```
Building targets
github.com/docker/docker/pkg/system
# github.com/docker/docker/pkg/system
pkg/mod/github.com/docker/docker@v0.7.3-0.20190309235953-33c3200e0d16/pkg/system/stat_linux.go:11:3: cannot use s.Rdev (type uint32) as type uint64 in field value
```

It seems this is identical to the docker one. IIRC there was a pull request to fix it, but it had no traction.

The @!mips line is acceptable.

**Comment by jmarcet** [comment]
> This is the error:
> 
> ```
> Building targets
> github.com/docker/docker/pkg/system
> # github.com/docker/docker/pkg/system
> pkg/mod/github.com/docker/docker@v0.7.3-0.20190309235953-33c3200e0d16/pkg/system/stat_linux.go:11:3: cannot use s.Rdev (type uint32) as type uint64 in field value
> ```
> 
> It seems this is identical to the docker one. IIRC there was a pull request to fix it, but it had no traction.
> 
> The @!mips line is acceptable.

OK, I reverted its drop.

**Comment by neheb** [comment]
ping @jeffreyto

File: utils/ctop/Makefile [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+PKG_RELEASE:=1
+PKG_LICENSE:=MIT
+PKG_LICENSE_FILES:=LICENSE
+
+PKG_VERSION:=0.7.3
+PKG_SOURCE:=v$(PKG_VERSION).tar.gz
+PKG_SOURCE_URL:=https://github.com/bcicen/ctop/archive
+PKG_HASH:=3b083b55a7cda7782e370bf03412dbf51ca4a94c1d56325ff70a1545d7a30adc
+
+PKG_MAINTAINER:=Javier Marcet <javier@marcet.info>
```
Comment (neheb) [inline utils/ctop/Makefile]: Place

PKG_LICENSE:=MIT
PKG_LICENSE_FILES:=LICENSE

below

**Comment by jefferyto** [comment]
> ping @jeffreyto

@neheb There is a slight misspelling here :joy:

**Comment by jefferyto** [comment]
Re: MIPS failure. Docker-/container-related packages appear to all have `@(aarch64||arm||x86_64)` (e.g. [docker-ce](https://github.com/openwrt/packages/blob/e2fe9bda3d6a2755f36031d8d69c6ea8e08ba715/utils/docker-ce/Makefile#L46), [runc](https://github.com/openwrt/packages/blob/e2fe9bda3d6a2755f36031d8d69c6ea8e08ba715/utils/runc/Makefile#L42)). I suspect containers are only available only on those platforms (for now at least). @G-M0N3Y-2503 may know more.

As I suspected in #12629, the CircleCI check [didn't actually run](https://app.circleci.com/pipelines/github/openwrt/packages/jobs/7545):

```
===+ Building: ctop skipped. It cannot be enabled with openwrt-sdk-ath79-generic_*.Linux-x86_64.tar.xz
```

I'm a little disappointed this PR was only compile-tested with one platform.

**Comment by neheb** [comment]
Well, docker runs mainly on those platforms. I assume MIPS platforms have too small RAM for docker to be feasible.

**Comment by G-M0N3Y-2503** [comment]
There were know bugs in some of dockers dependiencs preventing them to work on MIPS. It does look like docker is being patch more and more upstream to fix this, however doing so for OpenWRT (at least back then) involved porting a lot of patches. At the time it seemed like a lot less maintenance to wait for the upstream support.

However, given I have no MIPS hardware I can't test in either approach myself. I'm open to a PR, especially if it's not many patches now but also suspect that there aren't many MIPS users that would also want docker.

**Comment by neheb** [comment]
@G-M0N3Y-2503 you can test docker with the Malta target. Malta is MIPS for QEMU.

**Comment by G-M0N3Y-2503** [comment]
> MIPS for QEMU

hmmm, I guess I assumed emulation wouldn't cut it...
That said, for this issue, it looks like this package depends on an older version of docker, so I assume that this package would need patching for the dependency or depend on `docker-ce` somehow (assuming we get MIPS support for it)

**Pull request merged**